File: c++/2178.cpp
#include <iostream>
#include <queue>

using namespace std;
int N,M;
int map[101][101];
int miro[101][101];
bool visit[101][101];
int move_x[4]={0,1,-1,0};
int move_y[4]={1,0,0,-1};
queue<pair<int,int>> q;

void bfs(int x,int y){
    visit[x][y]=1;
    miro[x][y]=1;
    q.push({x,y});
    while(!q.empty()){
        int now_x=q.front().first;
        int now_y=q.front().second;
        q.pop();
        for(int i=0 ; i<4 ; i++){
            int next_x=now_x+move_x[i];
            int next_y=now_y+move_y[i];
            if(next_x>0 && next_x<=N && next_y>0 && next_y<=M){
                if(visit[next_x][next_y]==0 && map[next_x][next_y]==1){
                    q.push({next_x,next_y});
                    visit[next_x][next_y]=1;
                    miro[next_x][next_y]=miro[now_x][now_y]+1;
                }
            }
        }
    }
}

int main(){
    scanf("%d %d",&N,&M);
    for(int i=1 ; i<=N ; i++){
        for(int j=1 ; j<=M ; j++){
            scanf("%1d",&map[i][j]);
        }
    }
    bfs(1,1);
    printf("%d",miro[N][M]);
    return 0;
}

File: python/2178.py
N, M = map(int, input().split())
miro = []
for i in range(N):
    miro.append(list(map(int, input())))

finder = [[0 for i in range(M)] for j in range(N)]

move_x = [0, 1, 0, -1]
move_y = [1, 0, -1, 0]


def bfs(x, y):
    finder[x][y] = 1
    queue = []
    queue.insert(0, [x, y])
    while len(queue) > 0:
        now_x = queue[-1][0]
        now_y = queue[-1][1]
        queue.pop()
        for idx in range(4):
            next_x = now_x + move_x[idx]
            next_y = now_y + move_y[idx]
            if (next_x >= 0) and (next_y >= 0) and (next_x <= N-1) and (next_y <= M-1):
                if finder[next_x][next_y] == 0 and miro[next_x][next_y] == 1:
                    queue.insert(0, [next_x, next_y])
                    finder[next_x][next_y] = finder[now_x][now_y] + 1


bfs(0, 0)

print(finder[N-1][M-1])


